It drops support for Python3.9, it add support for Python 3.13, and it drops nox dependencies in favor of uv.

fressi-elastic opened the pull request:
In October 2025 Python.org is planning the release of Python 3.14 and dropping support for Python 3.9. This PR is following the trend dropping version 3.9 and adding the support of the now 1 year old Python 3.13.

It also remove testing Python 3.9 support in the `make test` goal. It instead provide goals to run tests with specific python versions (`test-3.10`, `test-3.11`, `test-3.12` and `test-3.13`, one for each supported version).
'make test' is now equivalent to 'make test-3.12`. 

The main idea is developers will test changes with their most convenient Python version, while the whole range of supported python versions should be better tested at continuous integration stage.

fressi-elastic committed:
Revert "Allow configuring the process startup method to be used for creating thespian actors. (#1975)"

This reverts commit 81abfc385aa27a785c43dfb98fddc58dddcefc9c.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally

fressi-elastic committed:
Add support for Python3.13 (pass unit tests).

fressi-elastic committed:
Revert unnecessary changes.

fressi-elastic committed:
Revert unnecessary changes.

fressi-elastic committed:
Revert unnecessary changes.

fressi-elastic committed:
Update the uv.lock file.

fressi-elastic committed:
Add entries for Python v3.13 to the CI testing matrix.

gareth-ellis reviewed (commented):
Can we update readme too? You mention that 3.12 is the preferred version, we should perhaps call that out in the readme too?

https://github.com/elastic/rally/blob/master/README.md?plain=1#L27

And also docs:
https://github.com/elastic/rally/blob/master/docs/install.rst?plain=1#L21

It may be worth grepping for 3.9, see if there's other occurances too

fressi-elastic committed:
It drops support for Python 3.9

fressi-elastic committed:
Fix docs/migrate.rst

fressi-elastic committed:
Remove statement: from __future__ import annotations

fressi-elastic committed:
Fix esrally/track/params.py and update uv.lock file.

fressi-elastic committed:
Revert changes to esrally/storage/_range.py

fressi-elastic committed:
Skip testing version 3.12 in buildkite pipeline.

fressi-elastic committed:
Fix some other annotation and one pylint founding.

gbanasiak commented on esrally/storage/_aws.py:
```diff
@@ -56,7 +54,7 @@ def __init__(
         aws_profile: str | None = AWS_PROFILE,
         chunk_size: int = CHUNK_SIZE,
-        s3_client: S3Client | None = None,
+        s3_client: Optional["S3Client"] = None,
```
`Optional` is not needed since 3.10, so why introducing it now?

gbanasiak replied:
Thanks for explaining, I'm good with what we have now :)

gbanasiak commented on esrally/storage/_transfer.py:
```diff
@@ -349,7 +348,7 @@ def close(self):
 
     @contextmanager
-    def _open(self) -> Iterator[FileDescriptor | None]:
+    def _open(self) -> Iterator[Optional["FileDescriptor"]]:
```
same here, is `Optional` required with 3.10?

gbanasiak commented on uv.lock:
```diff
@@ -12,9 +12,9 @@ resolution-markers = [
 version = "2.6.1"
 source = { registry = "https://pypi.org/simple" }
-sdist = { url = "https://files.pythonhosted.org/packages/26/30/f84a107a9c4331c14b2b586036f40965c128aa4fee4dda5d3d51cb14ad54/aiohappyeyeballs-2.6.1.tar.gz", hash = "sha256:c3f9d0113123803ccadfdf3f0faa505bc78e6a72d1cc4806cbd719826e943558", size = 22760, upload-time = "2025-03-12T01:42:48.764Z" }
+sdist = { url = "https://files.pythonhosted.org/packages/26/30/f84a107a9c4331c14b2b586036f40965c128aa4fee4dda5d3d51cb14ad54/aiohappyeyeballs-2.6.1.tar.gz", hash = "sha256:c3f9d0113123803ccadfdf3f0faa505bc78e6a72d1cc4806cbd719826e943558", size = 22760 }
```
I'm getting different results with the latest uv 0.8.22. I can see `upload-time` parameter in the updated lines. Can you run `uv self update` followed by `make uv-lock`?

gbanasiak reviewed (commented):
I think it would make sense to bump `DEFAULT_PY_VER` to 3.13 in this PR moving development environment to 3.13. If not now then when? Please also update Python versions in `mypy` and `black` configuration in `pyproject.toml`.

fressi-elastic commented:
Ok!

fressi-elastic committed:
Update CI python versions file.

fressi-elastic committed:
Remove unnecessary changes.

fressi-elastic committed:
Update CI python versions.

fressi-elastic committed:
Set Python3.13 as the default version.

fressi-elastic committed:
Let `make test` and `make it` use python 3.13

fressi-elastic committed:
Add standard-imghdr because imghdr has been removed from python 3.13

gareth-ellis commented on esrally/__init__.py:
```diff
@@ -72,9 +73,25 @@
+
+MIN_PYTHON_VERSION = Version(3, 10)
+
+MAX_PYTHON_VERSION = Version(3, 14)
```
Why not define this as 3,13 and then use >  in the comparison? It feels a little confusing (I assume its because you dont know what the max version may be....but that could be got around another way).

fressi-elastic replied:
I dropped the check for maxinum python version as it makes no sense to me. Anyway the information in `pyproject.toml` file should already help catching potential issues.

fressi-elastic committed:
Update docker image version.

fressi-elastic committed:
Use managed python version for preparing venv directory.

fressi-elastic committed:
Refactor Makefile:
- use uv instead of nox
- use .venv as deveult venv directory
- use .venv-$(PY_VERSION) when PY_VERSION is not empty.

fressi-elastic committed:
Add goal to run IT for serverless.

fressi-elastic committed:
Add goal for rally_tracks_compat

fressi-elastic committed:
Fix PHONY goals.

fressi-elastic committed:
Update python versions for mypy and black

fressi-elastic committed:
Update .buildkite pipeline files.

fressi-elastic committed:
Fix UV installation step.

fressi-elastic committed:
Debug CI scripts

fressi-elastic committed:
Fix Dockerfile and CI scripts.

fressi-elastic committed:
Fix VENV path in Dockerfile.

fressi-elastic committed:
Use levacy VIRTUAL_ENV as name for VENV_DIR

fressi-elastic committed:
Force Python version when running it_serverless test cases.

fressi-elastic committed:
Export PY_VERSION as an environment variable.

fressi-elastic committed:
Update uv version in Docerfile.

gbanasiak reviewed (commented):
Python upgrade PR turned into Python upgrade + Makefile refactor + nox replacement (although not complete as `noxfile.py` is still there). The title does not reflect the content. I suggest to reduce the scope to the original one in order to achieve the initial goal of enabling 3.13 while removing 3.9, and move nox removal to another PR. I don't feel sentimental about nox, if uv is faster let's do it.

fressi-elastic committed:
It drops the check for max python version.

fressi-elastic commented on .buildkite/it/run_serverless.sh:
```diff
@@ -17,22 +17,26 @@ export DEBIAN_FRONTEND=noninteractive
-"python${PYTHON_VERSION}" -m venv .venv
-source .venv/bin/activate
-pip install nox
+curl -LsSf https://astral.sh/uv/install.sh | sh
+source "${HOME}/.local/bin/env"
+
+echo "--- Create virtual environment"
+
+make venv
```
The test cases require a virtual env that was before provided by `nox`. UV is a replacement for it. It is required for running `make it_serverless`. The idea is letting things simpler and with lesser places to update every year we have a new python version to drop. This mean we don't need extra complexity and failure points like having to depend on `nox`.

gbanasiak replied:
Why https://github.com/elastic/ci-agent-images/blob/aec1570393a1c25fcb3240cefcb318128262964f/container-images/es-perf/python-base/Dockerfile#L11-L16 is not enough?

gbanasiak replied:
Right, `make venv` not only creates the venv but also populates it according to the uv lock (I think I assumed it's only venv creation earlier). But what about `curl -LsSf https://astral.sh/uv/install.sh | sh` ? It downloads and installs `uv` in Docker container where `uv` has already been installed. This seems unnecessary.

fressi-elastic committed:
Drop dependency on tox/nox.

fressi-elastic committed:
It ensures the rally plugin is installed before using for tests.

fressi-elastic committed:
It renames Makefile goal rally_tracks_compat -> it_tracks_compat

fressi-elastic committed:
It adds some comment in the Makefile.

fressi-elastic committed:
Remove unnecessary requirements and update wheel version.

gareth-ellis reviewed (approved):
LGTM

fressi-elastic committed:
Update Makefile error message.

fressi-elastic committed:
Update add_missing_loggers_to_config to make it a little easier to read.

gbanasiak commented on .buildkite/it/run_serverless.sh:
```diff
@@ -45,12 +49,13 @@ export THESPLOG_THRESHOLD="INFO"
 
 trap upload_logs ERR
 
+
 case $TEST_NAME in
     "user")
-        nox -s it_serverless
+        make -s it_serverless
```
`-s` in nox has different meaning than in `make`, is this intentional?

fressi-elastic replied:
It is not intentional! It's a mistake.

gbanasiak commented on Makefile:
```diff
@@ -16,105 +16,211 @@
+# It creates the project virtual environment.
+venv: uv $(VENV_ACTIVATE_FILE)
+
+$(VENV_ACTIVATE_FILE):
+	uv venv --allow-existing --seed
+	uv sync --locked --extra=develop
```
Any particular reason for reintroducing `pip` with `--seed`? Isn't the 2nd command enough?

gbanasiak commented on Makefile:
```diff
@@ -16,105 +16,211 @@
+install_pytest_rally_plugin: venv
+	$(VENV_ACTIVATE); uv pip install 'pytest-rally @ git+https://github.com/elastic/pytest-rally.git'
+
+# Old legacy alias goals
+install: venv
```
Previously `install` was unconditional which allowed going back and forth with the venv state when PR updates dependencies. Now once venv is installed, the only way to update dependencies is running `reinstall`. Is this intentional?

gbanasiak commented:
@pquentin Many thanks for comments.

> You didn't replace nox with uv, but with make, which is where the logic to test different Python versions exists.

Indeed.

> That said, if uv makes it easy to provide Python and install Rally now on Linux and macOS, maybe it's time to force a specific version and not deal with the overhead of supporting older versions.

`uv` definitely makes it easy for dev purposes, but how to ship Rally to end-users? Should we stop releasing PyPI packages and provide an installation script which uses `uv` and `git` or Docker under the hood? `curl -fsSL https://elastic.co/esrally | sh` ?
